src/backend/distributed/commands/multi_copy.c

```diff
@@ -2291,13 +2347,30 @@ CitusCopyDestReceiverStartup(DestReceiver *dest, int operation,
 	RecordRelationAccessIfNonDistTable(tableId, PLACEMENT_ACCESS_DML);
 
 	/*
-	 * For all the primary (e.g., writable) nodes, reserve a shared connection.
-	 * We do this upfront because we cannot know which nodes are going to be
-	 * accessed. Since the order of the reservation is important, we need to
-	 * do it right here. For the details on why the order important, see
-	 * the function.
+	 * For all the primary (e.g., writable) remote nodes, reserve a shared
+	 * connection. We do this upfront because we cannot know which nodes
+	 * are going to be accessed. Since the order of the reservation is
+	 * important, we need to do it right here. For the details on why the
+	 * order important, see the function.
+	 *
+	 * We don't need to case about local node because we either get a
+	 * connection or use local connection, so it cannot be part of
+	 * the starvation. As an edge case, we it cannot get a connection
+	 * and switch to local execution (e.g., disabled by user), COPY
+	 * would fail hinting the user to change the relevant settiing.
 	 */
-	EnsureConnectionPossibilityForPrimaryNodes();
+	EnsureConnectionPossibilityForRemotePrimaryNodes();
+
+	bool isIntermediateResult = copyDest->intermediateResultIdPrefix != NULL;
+	if (LocalCopyAllowed(isIntermediateResult))
+	{
+		/*
+		 * In this function, if needed, Citus may try to reserve a connection
+		 * to the local node as a side effect.
+		 */
+		copyDest->shouldUseLocalCopy = ShouldExecuteCopyLocally(shardIntervalList,
```

Comment:
is there a meaningful difference between LocalCopyAllowed and ShouldExecuteCopyLocally?